client/js/libs/handlebars/parse.js

```diff
@@ -1,126 +1,262 @@
 
-/**
- * MIRC compliant colour and style parser
- * Unfortuanately this is a non trivial operation
- * See this branch for source and tests
- * https://github.com/megawac/irc-style-parser/tree/shout
- */
-var styleCheck_Re = /[\x00-\x1F]/,
-	back_re = /^([0-9]{1,2})(,([0-9]{1,2}))?/,
-	colourKey = "\x03",
-	// breaks all open styles ^O (\x0F)
-	styleBreak = "\x0F";
+	var result = [];
 
+	var position = 0;
+	var start, end;
+	while (position < text.length) {
+		if (text[position] === ":" &&
+			text[position + 1] === "/" &&
+			text[position + 2] === "/") {
+			start = position;
+			// find scheme https://tools.ietf.org/html/rfc3986#section-3.1
+			var match = /[a-z][a-z0-9+-.]+$/i.exec(text.slice(0, start));
+			if (match) {
+				start = match.index;
+			}
+			end = position + 2;
+			while (end < text.length && /\S/.test(text[end])) {
+				end += 1;
+			}
+			if (match && end !== position + 2) {
+				while (punctuations.indexOf(text[start]) !== -1) {
+					start += 1;
+				}
+				while (punctuations.indexOf(text[end - 1]) !== -1) {
+					end -= 1;
+				}
+				result.push({
+					start: start,
+					end: end,
+					link: text.slice(start, end)
+				});
+				position = end;
+			}
+		} else if (text[position] === "w" &&
+			text[position + 1] === "w" &&
+			text[position + 2] === "w" &&
+			text[position + 3] === ".") {
+			start = position;
+			while (start > -1 && /\S/.test(text[start])) {
+				start -= 1;
+			}
+			start += 1;
 
-function styleTemplate(settings) {
-	return "<span class='" + settings.style + "'>" + settings.text + "</span>";
-}
+			end = position + 4;
+			while (end < text.length && /\S/.test(text[end])) {
+				end += 1;
+			}
 
-var styles = [
-	["normal", "\x00", ""], ["underline", "\x1F"],
-	["bold", "\x02"], ["italic", "\x1D"]
-].map(function(style) {
-	var escaped = encodeURI(style[1]).replace("%", "\\x");
-	return {
-		name: style[0],
-		style: style[2] ? style[2] : "irc-" + style[0],
-		key: style[1],
-		keyregex: new RegExp(escaped + "(.*?)(" + escaped + "|$)")
-	};
-});
-
-function colors(line) {
-	// http://www.mirc.com/colors.html
-	// http://www.aviran.org/stripremove-irc-client-control-characters/
-	// https://github.com/perl6/mu/blob/master/examples/rules/Grammar-IRC.pm
-	// regexs are cruel to parse this thing
-
-	// already done?
-	if (!styleCheck_Re.test(line)) {
-		return line;
-	}
+			if (end > position + 4) {
+				while (punctuations.indexOf(text[start]) !== -1) {
+					start += 1;
+				}
+				while (punctuations.indexOf(text[end - 1]) !== -1) {
+					end -= 1;
+				}
 
-	// split up by the irc style break character ^O
-	if (line.indexOf(styleBreak) >= 0) {
-		return line.split(styleBreak).map(colors).join("");
+				result.push({
+					start: start,
+					end: end,
+					link: "http://" + text.slice(start, end)
+				});
+			}
+		} else if ((text[position] === "#" || text[position === "&"]) &&
+			// ^-- this is basically wrong, because we need the 005 CHANTYPES response from the connect
```

Comment:
The front-end is actually aware of this, as I passed it into client when we moved to irc-fw. The network dom object has it in data.